File: SiliconLabsApp/BluetoothControllers/SILRSSIMeasurementTable.m
//
//  SILRSSIMeasurementTable.m
//  SiliconLabsApp
//
//  Created by Colden Prime on 2/6/15.
//  Copyright (c) 2015 SiliconLabs. All rights reserved.
//

#import "SILRSSIMeasurementTable.h"
#import "SILRSSIMeasurement.h"

@interface SILRSSIMeasurementTable ()

@property (strong, nonatomic) NSMutableArray *RSSIMeasurements;

@end

@implementation SILRSSIMeasurementTable

- (instancetype)init {
    self = [super init];
    if (self) {
        self.RSSIMeasurements = [NSMutableArray array];
    }
    return self;
}

- (void)addRSSIMeasurement:(NSNumber *)RSSI {
    NSInteger RSSIInteger = [RSSI integerValue];
    if (RSSIInteger <= 20 && RSSIInteger >= -100) {
        [self.RSSIMeasurements addObject:[[SILRSSIMeasurement alloc] initWithRSSI:RSSI date:[NSDate date]]];
    }
}

- (NSNumber *)lastRSSIMeasurement {
    return [[self.RSSIMeasurements lastObject] RSSI];
}

- (BOOL)hasRSSIMeasurementInPastTimeInterval:(NSTimeInterval)timeInterval {
    NSDate *referenceDate = [NSDate date];
    NSDate *lastRSSIMeasurementDate = [[self.RSSIMeasurements lastObject] date];
    return [referenceDate timeIntervalSinceDate:lastRSSIMeasurementDate] < timeInterval;
}

- (NSNumber *)averageRSSIMeasurementInPastTimeInterval:(NSTimeInterval)timeInterval {
    NSDate *date = [NSDate date];
    NSArray *filtered = [self.RSSIMeasurements filteredArrayUsingPredicate:[NSPredicate predicateWithBlock:^BOOL(SILRSSIMeasurement *measurement, NSDictionary *bindings) {
        NSDate *measurementDate = measurement.date;
        return [date timeIntervalSinceDate:measurementDate] < timeInterval;
    }]];

    if (filtered.count > 0) {
        return [filtered valueForKeyPath:@"@avg.RSSI.shortValue"];
    } else {
        return [NSNumber numberWithShort:0];
    }
}

@end

File: SiliconLabsApp/BluetoothControllers/SILRSSIMeasurementTable.swift
//
//  SILRSSIMeasurementTable.swift
//  BlueGecko
//
//  Created by Grzegorz Janosz on 09/03/2022.
//  Copyright © 2022 SiliconLabs. All rights reserved.
//

import Foundation
import RxCocoa
import RxSwift

struct SILRSSIMeasurement {
    let rssi: NSNumber
    let date: Date
}

@objcMembers
class SILRSSIMeasurementTable: NSObject {
    
    let rssiMeasurements = BehaviorRelay<[SILRSSIMeasurement]>(value: [])
    
    lazy var lastMeasurement: Observable<SILRSSIMeasurement> = rssiMeasurements.asObservable()
        .filter { !$0.isEmpty }
        .map { $0.last! }
    
    func addRSSIMeasurement(_ RSSI: NSNumber) {
        if -100...18 ~= RSSI.intValue {
            rssiMeasurements.addElement(element: SILRSSIMeasurement(rssi: RSSI, date: Date()))
        }
    }

    func lastRSSIMeasurement() -> NSNumber? {
        return rssiMeasurements.value.last?.rssi
    }

    func hasRSSIMeasurement(inPastTimeInterval timeInterval: TimeInterval) -> Bool {
        guard let lastRSSIMeasurementDate = rssiMeasurements.value.last?.date else {
            return false
        }
        return Date().timeIntervalSince(lastRSSIMeasurementDate) < timeInterval
    }

    func averageRSSIMeasurement(inPastTimeInterval timeInterval: TimeInterval) -> NSNumber {
        let filtered = rssiMeasurements.value.filter { Date().timeIntervalSince($0.date) < timeInterval }
        
        if filtered.count > 0 {
            let rssiSum = filtered.reduce(0) { $0 + $1.rssi.intValue }
            return NSNumber(value: Double(rssiSum) / Double(filtered.count))
        } else {
            return NSNumber(value: 0)
        }
    }
}
